projects/scion/workbench-application-platform/src/lib/core/qualifier-patcher.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2018-2019 Swiss Federal Railways
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ *  SPDX-License-Identifier: EPL-2.0
+ */
+
+import { NilQualifier, Qualifier } from '@scion/workbench-application-platform.api';
+
+/**
+ * Returns a copy of the given intent qualifier with all its wildcard qualifier values
+ * replaced with values of the given capability qualifier, if any.
+ * Also, if the intent qualifier specifies a wildcard key, it is merged with the capability qualifier.
+ *
+ * @param intentQualifier
+ *        qualifier for an intent as specified in the manifest, may contain wildcards as qualifier key (*)
+ *        and/or qualifier value (* or ?).
+ * @param capabilityQualifier
+ *        qualifier for a capability as specified in the manifest, may contain wildcards (* or ?) as qualifier value;
+ *        if `null`, {NilQualifier} is used.
+ */
+export function patchQualifier(intentQualifier: Qualifier, capabilityQualifier: Qualifier): Qualifier {
+  const _capabilityQualifier = capabilityQualifier || NilQualifier;
+  const _intentQualifier = intentQualifier || NilQualifier;
+
+  // Create a working copy of the intent qualifier
+  const _patchedQualifier: Qualifier = Object.assign({}, _intentQualifier);
+  const isAnyQualifier = _patchedQualifier.hasOwnProperty('*');
+  delete _patchedQualifier['*'];
+
+  Object.keys(_capabilityQualifier)
+    .forEach(key => {
+      if (_intentQualifier[key] === '*' || _intentQualifier[key] === '?') {
+        _patchedQualifier[key] = _capabilityQualifier[key];
+      } else if (isAnyQualifier && !_patchedQualifier.hasOwnProperty(key)) {
```

Comment:
Use `_intentQualifier` instead of `_patchedQualifier` (easier to understand)

```
} else if (isAnyQualifier && !_intentQualifier.hasOwnProperty(key)) {
```

Reply:
Done